homeassistant/components/tplink/entity.py

```diff
@@ -62,24 +108,278 @@ async def _async_wrap(self: _T, *args: _P.args, **kwargs: _P.kwargs) -> None:
+
+        return entity_category
+
+    @classmethod
+    def _description_for_feature[_D: EntityDescription](
+        cls,
+        desc_cls: type[_D],
+        feature: Feature,
+        descriptions: Mapping[str, _D],
+        *,
+        child_alias: str | None = None,
+    ) -> _D:
+        """Return description object for the given feature.
+
+        This is responsible for setting the common parameters & deciding based on feature id
+        which additional parameters are passed.
+        """
+
+        enabled_default = feature.category is not Feature.Category.Debug
+        category = cls._category_for_feature(feature)
+        if descriptions and (desc := descriptions.get(feature.id)):
+            translation_key: str | None = feature.id
+            # HA logic is to name entities based on the following logic:
+            # _attr_name > translation.name > description.name
+            # > device_class (if base platform supports).
+            entity_name: str | None | UndefinedType = UNDEFINED
+
+            # The state feature gets the device name or the child device
+            # name if it's a child device
+            if feature.id == PRIMARY_STATE_ID:
+                translation_key = None
+                entity_name = child_alias  # if None will use device name
+
+            desc = replace(
+                desc,
+                translation_key=translation_key,
+                name=entity_name,  # if undefined will use translation key
+                entity_category=category,
+                entity_registry_enabled_default=enabled_default,
+            )
+        else:
+            desc = desc_cls(
+                key=feature.id,
+                name=feature.name,
+                icon=feature.icon,
+                entity_category=category,
+                entity_registry_enabled_default=enabled_default,
+            )
+            _LOGGER.warning(
+                "Device feature: %s (%s) needs an entity description defined, "
+                "it has been added to hass but may not be presented properly",
+                feature.name,
+                feature.id,
+            )
+        return desc
+
+
+def _entities_for_device[_E: CoordinatedTPLinkEntity](
+    device: Device,
+    coordinator: TPLinkDataUpdateCoordinator,
+    *,
+    feature_type: Feature.Type,
+    entity_class: type[_E],
+    parent: Device | None = None,
+) -> list[_E]:
+    """Return a list of entities to add.
+
+    This filters out unwanted features to avoid creating unnecessary entities
+    for device features that are implemented by specialized platforms like light.
+    """
+    entities: list[_E] = [
+        entity_class(
+            device,
+            coordinator,
+            feature=feat,
+            parent=parent,
+        )
+        for feat in device.features.values()
+        if feat.type == feature_type
+        and (
+            feat.category != Feature.Category.Primary
```

Comment:
```suggestion
            feat.category is not Feature.Category.Primary
```

Reply:
Fixed in https://github.com/home-assistant/core/pull/120141